tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs

```diff
@@ -0,0 +1,359 @@
+                self.tail += 1;
+            },
+            (true, false, _) => unsafe {
+                self.copy(idx, idx + 1, self.head - idx - 1);
+                self.head -= 1;
+            },
+            (false, true, true) => unsafe {
+                self.copy(self.tail + 1, self.tail, index);
+                self.tail = self.wrap_add(self.tail, 1);
+            },
+            (false, false, false) => unsafe {
+                self.copy(idx, idx + 1, self.head - idx - 1);
+                self.head -= 1;
+            },
+            (false, false, true) => {
+                unsafe {
+                    // draw in elements in the tail section
+                    self.copy(idx, idx + 1, self.cap() - idx - 1);
+
+                    // Prevents underflow.
+                    if self.head != 0 {
+                        // copy first element into empty spot
+                        self.copy(self.cap() - 1, 0, 1);
+
+                        // move elements in the head section backwards
+                        self.copy(0, 1, self.head - 1);
+                    }
+
+                    self.head = self.wrap_sub(self.head, 1);
+                }
+            }
+            (false, true, false) => {
+                unsafe {
+                    // draw in elements up to idx
+                    self.copy(1, 0, idx);
+
+                    // copy last element into empty spot
+                    self.copy(0, self.cap() - 1, 1);
+
+                    // move elements from tail to end forward, excluding the last one
+                    self.copy(self.tail + 1, self.tail, self.cap() - self.tail - 1);
+
+                    self.tail = self.wrap_add(self.tail, 1);
+                }
+            }
+        }
+
+        elem
+    }
+
+    // Kani change: abstract by not doing any copies
+    unsafe fn copy(&self, dst: usize, src: usize, len: usize) {
+        assert!(dst + len <= self.cap());
+        assert!(src + len <= self.cap());
+    }
+
+    fn wrap_index(&self, idx: usize) -> usize {
+        wrap_index(idx, self.cap())
+    }
+
+    fn wrap_add(&self, idx: usize, addend: usize) -> usize {
+        wrap_index(idx.wrapping_add(addend), self.cap())
+    }
+
+    fn wrap_sub(&self, idx: usize, subtrahend: usize) -> usize {
+        wrap_index(idx.wrapping_sub(subtrahend), self.cap())
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.tail == self.head
+    }
+
+    fn is_contiguous(&self) -> bool {
+        self.tail <= self.head
+    }
+}
+
+fn wrap_index(index: usize, size: usize) -> usize {
+    // size is always a power of 2
+    assert!(size.is_power_of_two());
+    index & (size - 1)
+}
+
+fn count(tail: usize, head: usize, size: usize) -> usize {
+    // size is always a power of 2
+    assert!(is_nonzero_pow2(size));
+    (head.wrapping_sub(tail)) & (size - 1)
+}
+
+///
+/// Based on src/alloc/raw_vec.rs
+///
+/// Generic type `T` is implicit (but we assume it is not a ZST)
+struct AbstractRawVec {
```

Comment:
Is there a trait bound for `Not ZST`?

Reply:
Seems not: https://users.rust-lang.org/t/how-to-mark-non-zst/74842